CB-11530 Upload freeipa-client dependencies to nexus

kmanamcheri opened the pull request:
Previously, in CDPSDX-2788, the build scripts were updated to upload
freeipa-client as a jar into nexus. However, freeipa-client has two
dependencies: common, and cluster-proxy. We need to upload these
also to nexus so that thunderhead can use the freeipa-client.

Verified that common and cluster-proxy do not have any further
dependencies.

See detailed description in the commit message.

jamisonbennett commented:
I would like to see if we can remove the freeipa client dependency on the cluster proxy. I would think it would be best if we could inject an interface that handles this. That way other consumers that don't know anything about cluster proxy don't need to have the dependency.

lacikaaa commented:
I think both cluster-proxy and common module has spring-boot dependency so you should use the `uploadBootArchives` task to upload them.
Also I don't know how much is used from these modules. If only a few class maybe it worth duplicating them instead of shipping those with freeipa-client

kmanamcheri commented:
> I think both cluster-proxy and common module has spring-boot dependency so you should use the `uploadBootArchives` task to upload them.
> Also I don't know how much is used from these modules. If only a few class maybe it worth duplicating them instead of shipping those with freeipa-client

Good point. I'll address both yours and Jamison's comment by trying to remove the dependencies by duplicating them or making them a runtime dependency.

kmanamcheri commented:
@jamisonbennett I removed the cluster proxy dependency by making the exception classes as interfaces and abstract. This is probably ok assuming that `thunderhead` has its own ClusterProxyException and ClusterProxyError classes.

@lacikaaa I could not remove the `common` dependency because it pulls in a bunch of other things (like Tracer, common-logging framework etc). So I think its best to just get the common module as a dependency as well. I modified to use `uploadBootArchives`

mahendra-rk reviewed (approved):
LGTM!

lacikaaa commented on build.gradle:
```diff
@@ -232,7 +232,7 @@ subprojects {
       }
     }
-    if (project.name in ['freeipa-client']) {
+    if (project.name in ['freeipa-client', 'common']) {
```
I think `common` should be added to line 228

lacikaaa commented on scripts/build-dev.sh:
```diff
@@ -1,7 +1,7 @@
 set -x
 
-./gradlew -Penv=jenkins -b build.gradle buildInfo build uploadBootArchives :freeipa-client:uploadArchives -Pversion=$VERSION --parallel --stacktrace -x checkstyleMain -x checkstyleTest -x spotbugsMain -x spotbugsTest
+./gradlew -Penv=jenkins -b build.gradle buildInfo build uploadBootArchives :freeipa-client:uploadArchives :common:uploadBootArchives -Pversion=$VERSION --parallel --stacktrace -x checkstyleMain -x checkstyleTest -x spotbugsMain -x spotbugsTest
```
I don't think you have to add this common related task. the same for `build-release.sh`

kmanamcheri replied:
Is it because there is the 'uploadBootArchives' which runs for all projects?

lacikaaa replied:
yes

lacikaaa commented:
have you managed to test:
- the upload works for common? (eg with using local nexus)
- `ClusterProxyException` handling in the client still works as before?
